### ISSUE TITLE
Expression interpreter can avoid allocating (empty) arrays

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/corefx#10520]
Did a little performance tracing session for simple expressions evaluated using the interpreter. Not surprising, a big source of allocations is for `object[]` instances passed to `Invoke` methods in reflection. I'm opening this issue to track some possible improvements in this area.

Instructions for `Call`, `New`, etc. seem to allocate an empty `object[]` in case there are no arguments, in order to call the `Invoke` method through reflection. We should be able to avoid these allocations here and pass `Array.Empty<object>()` instead.

~~In addition, we may be able to reduce `object[]` allocations in general, given that reflection also copies the argument array in `CheckArguments`, by caching a limited set of reusable `object[]` buffers for small lengths. I don't think `System.Buffers` fits the bill unfortunately, because it will hand out buffers of a minimum size or larger. Unless reflection would be able to operate on `ArraySegment<object>` to 'trim' the view over a rented array, we'd have to keep our own cache. An obvious question is the trade-off between allocating arrays and the (synchronization) cost of renting them from a pool.~~

**Comment by JonHanna** [comment]
The trade-off for non-zero-length arrays could be complicated to judge with confidence and need some testing. The trade-off for using Array.Empty when applicable seems more straight-forward, so perhaps start with just that.

**Comment by bartdesmet** [comment]
Yup, prepared just that. In fact, I've been able to get rid of another copy created for instance method calls, where we were skipping over the instance operand and created a brand new array for the arguments. I'm working on trying to add a performance test project but fighting a build error. When I get over that I'll send out a PR.

**Comment by bartdesmet** [comment]
Initial results look promising:

| Kind | Args | Avg 100x before (ms) | Avg 100x after (ms) | Ratio after/before |
| ---- | ---- | ---- | ---- | ---- | ---- |
| Static | 0 | 0.021 | 0.019 |  0.905 |
| Static | 1 | 0.031 | 0.029 | 0.935 |
| Static | 2 | 0.040 | 0.037 | 0.925 |
| Static | 3 | 0.050 | 0.047 | 0.940 |
| Static | 4 | 0.058 | 0.054 | 0.931 |
| Instance | 0 | 0.025 | 0.021 | 0.840 |
| Instance | 1 | 0.038 | 0.031 | 0.816 |
| Instance | 2 | 0.048 | 0.039 | 0.813 |
| Instance | 3 | 0.056 | 0.05 | 0.893 |
| Instance | 4 | 0.067 | 0.057 | 0.851 |

The test is effectively calling a static or instance method with the specified number of arguments a 100 times within a loop formulated using an expression tree:

```csharp
for (var i = 0; i < 100; I++)
    call(...);
```

These numbers are obtained by just applying three minor optimizations:

* Reducing the number of virtual method indirections in `MethodInfoCallInstruction`
* Using `Array.Empty<object>()` in case of calling a method with no parameters
* Avoiding one arguments array copy for instance methods (cf. `SkipOne` method)

Note that `MethodBase.Invoke` is also used in the interpreter to execute property and indexer accessors, so these gain as well (especially instance properties, which no longer suffer from two redundant `object[]` allocations).

The test suite looks as follows:

```csharp
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;

class Program
{
    static void Main()
    {
        var M = 100;
        var N = 100000;

        for (var i = 0; i <= 4; i++)
        {
            var m = typeof(Static).GetMethod("M" + i);
            var a = Enumerable.Range(0, i).Select(j => Expression.Constant(j)).ToArray();
            var c = Expression.Call(m, a);
            var e = GetLoop(c, M);
            Test(e, true, N, c.ToString());
        }

        for (var i = 0; i <= 4; i++)
        {
            var o = Expression.Constant(new Instance());
            var m = typeof(Instance).GetMethod("M" + i);
            var a = Enumerable.Range(0, i).Select(j => Expression.Constant(j)).ToArray();
            var c = Expression.Call(o, m, a);
            var e = GetLoop(c, M);
            Test(e, true, N, c.ToString());
        }
    }

    static Expression<Action> GetLoop(Expression e, int M)
    {
        var i = Expression.Parameter(typeof(int));
        var b = Expression.Label();
        return
            Expression.Lambda<Action>(
                Expression.Block(
                    new[] { i },
                    Expression.Assign(i, Expression.Constant(0)),
                    Expression.Loop(
                        Expression.Block(
                            Expression.IfThen(
                                Expression.GreaterThan(
                                    Expression.PostIncrementAssign(i),
                                    Expression.Constant(M)
                                ),
                                Expression.Break(b)
                            ),
                            e
                        ),
                        b
                    )
                )
            );
    }

    static void Test(Expression<Action> e, bool interpret, int N, string title = null)
    {
        var f = e.Compile(interpret);

        var times = new List<double>(N);

        var sw = new Stopwatch();
        for (var i = 0; i < N; i++)
        {
            sw.Restart();
            f();
            times.Add(sw.Elapsed.TotalMilliseconds);
        }

        times.Sort();

        var s = (title ?? e.ToString()) + (interpret ? " - Interpreted" : " - Compiled");
        Console.WriteLine(s);
        Console.WriteLine(new string('-', s.Length));
        Console.WriteLine("Avg: {0:F6}", times.Average());
        Console.WriteLine("Min: {0:F6}", times[0]);
        Console.WriteLine("50%: {0:F6}", times[(int)(N * 0.50)]);
        Console.WriteLine("90%: {0:F6}", times[(int)(N * 0.90)]);
        Console.WriteLine("95%: {0:F6}", times[(int)(N * 0.95)]);
        Console.WriteLine("99%: {0:F6}", times[(int)(N * 0.99)]);
        Console.WriteLine("Max: {0:F6}", times[N - 1]);
        Console.WriteLine();
    }
}

class Static
{
    public static int M0() => 42;
    public static int M1(int p1) => 42;
    public static int M2(int p1, int p2) => 42;
    public static int M3(int p1, int p2, int p3) => 42;
    public static int M4(int p1, int p2, int p3, int p4) => 42;
}

class Instance
{
    public int M0() => 42;
    public int M1(int p1) => 42;
    public int M2(int p1, int p2) => 42;
    public int M3(int p1, int p2, int p3) => 42;
    public int M4(int p1, int p2, int p3, int p4) => 42;
}
```

**Comment by bartdesmet** [comment]
Pushed the low hanging fruit changes. Some initial experimentation for non-zero-length arrays passed to `Invoke` doesn't add much value unless the outgoing call is significantly heavy and causes the freshly allocated `object[]` to age beyond gen 0. For short-lived calls, the overhead causes performance degradations. Given that the invocation target is a black box to us, there's no good way to have a generally applicable enhancement at this level.

A better candidate for reflection optimization (for our use case, or more generally) may be in reflection itself. We already noted that a while ago when we concluded changes were needed there to support e.g. passing `Type.Missing` in a verbatim manner. Also, some of the safety checks there are redundant given that we perform similar checks in the factories and/or compiler. Finally, for our use case, the array does not have to be copied by reflection; we don't mutate it after passing it to `Invoke`.
